File: lua/src/sekreto/plugins/infisical.lua
-- The infisical plugin: Infisical, over HTTPS, with a token or a
-- universal-auth machine identity.
--
-- A port of typescript/plugins/infisical.ts, which is canonical.

local err = require('sekreto.err')
local name = require('sekreto.name')
local addr = require('sekreto.addr')
local providers = require('sekreto.providers')
local json = require('sekreto.plugins.json')
local httpjson = require('sekreto.plugins.httpjson')
local support = require('sekreto.plugins.support')

local fail = err.fail
local nonempty = providers.nonempty
local trimslash = providers.trimslash
local providerplugin = providers.providerplugin
local first = support.first
local nowms = support.nowms
local renewtime = support.renewtime
local NEVER = support.NEVER
local fetchjson = httpjson.fetchjson
local uriescape = support.uriescape

local M = {}

--- Infisical.
---
--- `api.token` reads the secret keyed `API_TOKEN` at a secret path in one
--- environment of a project. Auth is a token, or a universal-auth
--- (machine identity) login with clientid/clientsecret.
local function infisical(useaddr, token, clientid, clientsecret, project, environment, path)
  local livetoken = nil
  local renewat = NEVER

  local function login(address)
    if nonempty(token) then
      return token
    end

    if not nonempty(clientid) or not nonempty(clientsecret) then
      fail('sekreto: infisical: no token and no client credentials')
    end

    local body = json.obj({
      { 'clientId', clientid },
      { 'clientSecret', clientsecret },
    })

    local res = fetchjson(
      'POST', address .. '/api/v1/auth/universal-auth/login',
      { { 'content-type', 'application/json' } }, json.stringify(body)
    )

    local got = json.text(json.dig(res.body, 'accessToken'))
    if 200 ~= res.status or nil == got or '' == got then
      fail('sekreto: infisical login failed: ' .. res.status)
    end

    -- camelCase, unlike everyone else's expires_in.
    renewat = renewtime(json.dig(res.body, 'expiresIn'))

    return got
  end

  return {
    lookup = function(secret)
      local address = trimslash(first(useaddr, 'https://app.infisical.com'))
      addr.checkaddr(address)

      local useproject = project or ''
      local useenvironment = environment or ''
      if '' == useproject or '' == useenvironment then
        fail('sekreto: infisical: no project/environment')
      end

      if nil == livetoken or nowms() >= renewat then
        livetoken = login(address)
      end

      local url = address .. '/api/v3/secrets/raw/' .. name.envkey(secret) ..
        '?workspaceId=' .. uriescape(useproject) ..
        '&environment=' .. uriescape(useenvironment) ..
        '&secretPath=' .. uriescape(first(path, '/'))

      local res = fetchjson('GET', url, { { 'authorization', 'Bearer ' .. livetoken } })

      if 404 == res.status then
        return nil
      end

      if 200 ~= res.status then
        fail('sekreto: infisical error: ' .. res.status)
      end

      return json.text(json.dig(res.body, 'secret', 'secretValue'))
    end,
    describe = function()
      return 'infisical:' .. (project or '') .. '/' .. (environment or '')
    end,
  }
end

--- The kind, as a voxgig/plugin definition.
M.infisical = providerplugin('infisical', function(spec)
  return infisical(spec.addr, spec.token, spec.clientid, spec.clientsecret,
    spec.project, spec.environment, spec.path)
end)

return M

File: java/plugins/com/voxgig/sekreto/plugins/Infisical.java
// Infisical, as a voxgig/plugin definition.
//
// A PLUGIN, NOT PART OF THE CORE: it opens a socket, and the core reads
// at most a local file. A chain that does not name this kind must not
// carry it, so the calling project imports this definition and hands it
// to Sekreto in the `plugins` option.
// See docs/design/plugin-providers.md.

package com.voxgig.sekreto.plugins;

import static com.voxgig.sekreto.Addr.checkaddr;
import static com.voxgig.sekreto.plugins.Httpjson.dig;
import static com.voxgig.sekreto.plugins.Httpjson.fetchjson;
import static com.voxgig.sekreto.plugins.Httpjson.first;
import static com.voxgig.sekreto.plugins.Httpjson.headers;
import static com.voxgig.sekreto.plugins.Httpjson.renewtime;
import static com.voxgig.sekreto.plugins.Httpjson.trimslash;

import com.voxgig.sekreto.Json;
import com.voxgig.sekreto.Provider;
import com.voxgig.sekreto.Sekreto;
import com.voxgig.sekreto.Sekreto.SekretoError;
import com.voxgig.sekreto.Support;
import com.voxgig.sekreto.plugins.Httpjson.Answer;
import java.util.LinkedHashMap;
import java.util.Map;
import voxgig.plugin.Definition;

/**
 * Infisical.
 *
 * <p>`api.token` reads the secret keyed `API_TOKEN` (Infisical's own
 * convention is environment-style keys) at a secret path in one
 * environment of a project. Auth is a token, or a universal-auth (machine
 * identity) login with clientid/clientsecret.
 */
public final class Infisical implements Provider {

  /** The `infisical` kind: what `plugins` hands to Sekreto. */
  public static final Definition PLUGIN = Support.providerplugin("infisical", spec ->
      new Infisical(
          Support.text(spec.get("addr")), Support.text(spec.get("token")),
          Support.text(spec.get("clientid")), Support.text(spec.get("clientsecret")),
          Support.text(spec.get("project")), Support.text(spec.get("environment")),
          Support.text(spec.get("path"))));
  private final String addr;
  private final String token;
  private final String clientid;
  private final String clientsecret;
  private final String project;
  private final String environment;
  private final String path;

  // A configured token is kept forever; a universal-auth token carries
  // expiresIn and is renewed shortly before it runs out.
  private String livetoken;
  private long renewat = Long.MAX_VALUE;

  public Infisical(String addr, String token, String clientid, String clientsecret,
      String project, String environment, String path) {
    this.addr = addr;
    this.token = token;
    this.clientid = clientid;
    this.clientsecret = clientsecret;
    this.project = project;
    this.environment = environment;
    this.path = path;
  }

  private String login(String useaddr) {
    if (null != token && !token.isEmpty()) {
      return token;
    }

    if (null == clientid || clientid.isEmpty()
        || null == clientsecret || clientsecret.isEmpty()) {
      throw new SekretoError("sekreto: infisical: no token and no client credentials");
    }

    Map<String, Object> body = new LinkedHashMap<>();
    body.put("clientId", clientid);
    body.put("clientSecret", clientsecret);

    Answer res = fetchjson("POST", useaddr + "/api/v1/auth/universal-auth/login",
        headers("content-type", "application/json"), Json.stringify(body));

    Object got = dig(res.body, "accessToken");
    if (200 != res.status || null == got || String.valueOf(got).isEmpty()) {
      throw new SekretoError("sekreto: infisical login failed: " + res.status);
    }

    renewat = renewtime(dig(res.body, "expiresIn"));

    return String.valueOf(got);
  }

  @Override
  public String lookup(String name) {
    String useaddr = trimslash(first(addr, "https://app.infisical.com"));
    checkaddr(useaddr);

    String useproject = null == project ? "" : project;
    String useenvironment = null == environment ? "" : environment;
    if (useproject.isEmpty() || useenvironment.isEmpty()) {
      throw new SekretoError("sekreto: infisical: no project/environment");
    }

    if (null == livetoken || System.currentTimeMillis() >= renewat) {
      livetoken = login(useaddr);
    }

    String url = useaddr + "/api/v3/secrets/raw/" + Sekreto.envkey(name, null)
        + "?workspaceId=" + Sigv4.uriescape(useproject)
        + "&environment=" + Sigv4.uriescape(useenvironment)
        + "&secretPath=" + Sigv4.uriescape(first(path, "/"));

    Answer res = fetchjson("GET", url, headers("authorization", "Bearer " + livetoken), null);

    if (404 == res.status) {
      return null;
    }

    if (200 != res.status) {
      throw new SekretoError("sekreto: infisical error: " + res.status);
    }

    Object value = dig(res.body, "secret", "secretValue");
    return null == value ? null : String.valueOf(value);
  }

  @Override
  public String describe() {
    return "infisical:" + (null == project ? "" : project)
        + "/" + (null == environment ? "" : environment);
  }
}
